src/tools/tools_execute.rs

```diff
@@ -232,9 +232,25 @@ async fn pp_run_tools(
                 "context_file".to_string(),
                 serde_json::to_string(&json_vec).unwrap()
             );
-            generated_other.push(message.clone());
+            let mut found_exact_same_message = false;
+            for original_msg in original_messages.iter().rev() {
+                if original_msg.role == "user" {
+                    break;
+                }
+                if original_msg.content == message.content {
+                    found_exact_same_message = true;
+                    break;
+                }
+            }
+            if !found_exact_same_message {
+                generated_other.push(message.clone());
+            } else {
+                generated_other.push(ChatMessage::new(
```

Comment:
isn't it supposed to be role="tool" message? So model would know, this specific tool call caused it, not a random one